### ISSUE TITLE
[intern-18] Audio decoding & Whisper model loading: trace pipeline

### ISSUE THREAD

**Issue by anhnth24** [anhnth24/project-example#356]
## Mục tiêu học
Hiểu cách backend load & cache Whisper model; decode audio (mp3/wav/ogg) → 16kHz; transcode bằng whisper-rs.

## Scope
`crates/core/src/audio.rs` (read-only). Không train model hay optimize inference.

## Techstack / khái niệm
- `AudioEngine` + `LoadOnceCache` (process-wide LRU for models)
- `symphonia` decoder → `rubato` resampler (FFT-based)
- Whisper language = "vi" (Vietnamese)
- `no_speech_probability` filtering
- PhoWhisper (VinAI fine-tune) vs standard whisper-small

## Đọc theo thứ tự
1. `CLAUDE.md` mục kiến trúc → `audio.rs` phần cache + load
2. `docs/codebase-summary.md` — Audio conversion detail
3. `crates/core/src/audio.rs` lines 1–200 (cache + loading)
4. `bench/download_models.sh` — model files location

## Cách làm
1. Đảm bảo có Whisper model (tiny hoặc base từ `bench/download_models.sh`).
2. Chạy `./target/release/fileconv audio tiny <manifest.tsv> [out.md]` trên test audio nhỏ.
3. Trace code: cache hit/miss logic? resample xảy ra ở đâu? Language code "vi" được sử dụng sao?
4. Ghi: "LRU eviction kích hoạt khi nào?" + "Tại sao dùng rubato FFT thay vì linear?"

## Acceptance criteria
- [ ] Chạy audio convert thành công (hoặc skip nếu không có model)
- [ ] Giải thích được: LoadOnceCache pattern + TTL/capacity
- [ ] Mô tả resampling pipeline (decode → resample → whisper)

## Không làm
- Không train/download model bằng tay (dùng script sẵn)
- Không thay `no_speech_probability` threshold

## Gợi ý PR
Branch: `intern/18-trace-audio-pipeline`.

**Comment by Supermonters** [comment]
Anh @anhnth24 ơi em đã làm xong issue #356 và em tóm lại luồng audio/Whisper như sau ạ.

Issue này chỉ cần đọc code + chạy thử, nên em **không sửa code**, không đổi `no_speech` threshold, không train model, cũng **không mở PR/branch** ạ.

## Lệnh đã chạy

```bash
cargo build --release -p fileconv-cli
bash bench/download_models.sh
# (macOS: dòng `stat -c%s` trong script báo lỗi GNU-stat, nhưng curl vẫn tải file OK)
python3 -m venv .venv && source .venv/bin/activate
pip install gTTS
python3 bench/make_vn_audio.py   # clip1–4.mp3 + .txt trong bench/vn_audio

./target/release/fileconv audio models/ggml-tiny.bin \
  bench/vn_audio/manifest.tsv /tmp/intern18-audio-tiny.md

./target/release/fileconv audio models/ggml-tiny.bin,models/ggml-PhoWhisper-small.bin \
  bench/vn_audio/manifest.tsv /tmp/intern18-compare.md

./target/release/fileconv one bench/vn_audio/clip1.mp3
# (tuỳ chọn) ép tiny:
# FILECONV_WHISPER_MODEL=models/ggml-tiny.bin ./target/release/fileconv one bench/vn_audio/clip1.mp3

cargo test -p fileconv-core --features audio load_once_cache_
```

Model có sẵn: `ggml-tiny` 74M, `base` 141M, `small` 465M, `ggml-PhoWhisper-small` 465M.

Ground truth `clip1.txt` (gTTS):
`Xin chào, hôm nay tôi thử nghiệm công cụ chuyển giọng nói thành văn bản bằng tiếng Việt.`

## Kết quả

### `fileconv audio` — ggml-tiny (load 1 lần ~75ms rồi cache)

| Model | Clip | Acc | CER | WER | RTF |
|-------|------|-----|-----|-----|-----|
| ggml-tiny | clip1 | **93.2%** | 0.068 | 0.111 | 0.03 |
| ggml-tiny | clip2 | **96.5%** | 0.035 | 0.095 | 0.02 |
| ggml-tiny | clip3 | **90.5%** | 0.095 | 0.235 | 0.03 |
| ggml-tiny | clip4 | **77.1%** | 0.229 | 0.381 | 0.02 |

- Trung bình tiny: **89.3%** accuracy, WER **0.206**, RTF **0.03**
- Report ghi rõ: model **load 1 lần rồi cache** — khớp `LoadOnceCache`

Transcript tiny (ép `FILECONV_WHISPER_MODEL=…/ggml-tiny.bin`) gần GT, còn lệch nhẹ kiểu *công cuộc* / *tiếng viên*.

### So sánh tiny vs PhoWhisper (cùng set gTTS)

| Model | Clip1 | Clip2 | Clip3 | Clip4 | Avg |
|-------|-------|-------|-------|-------|-----|
| ggml-tiny | 93.2% | 96.5% | 90.5% | 77.1% | **89.3%** |
| PhoWhisper-small | **1.1%** | 87.1% | **9.5%** | 89.2% | **46.7%** |

→ Trên corpus **gTTS synthetic** này PhoWhisper **không** tốt hơn tiny (clip1/clip3 gần như fail). Em không kết luận PhoWhisper kém trên giọng người thật — chỉ ghi nhận lệch domain với gTTS / discovery ưu tiên PhoWhisper.

### `fileconv one clip1.mp3` (không set env)

Output chỉ còn `.` — vì `discover_whisper_model()` ưu tiên `ggml-PhoWhisper-small.bin` trước small/base/tiny. Khả năng cao: segment bị lọc (`no_speech_thold=0.6` + filter marker) hoặc ảo giác còn sót `.`. Em **không** đụng threshold theo scope issue.

### Unit test

`cargo test -p fileconv-core --features audio load_once_cache_` → **12 passed**.

## Em hiểu pipeline như thế này

1. `FormatKind::from_path` → đuôi `.mp3/.wav/...` → `FormatKind::Audio` (không sniff magic).
2. `Converter::engine()` lấy model: `opts.whisper_model` hoặc `discover_whisper_model()` (PhoWhisper → small → base → tiny).
3. `AudioEngine::load` → `LoadOnceCache` process-wide lấy `Arc<WhisperContext>`.
4. Decode: `symphonia` → mono (average channels) → `rubato` FFT resample **16 kHz** → trim `output_delay`.
5. Whisper `full` với lang mặc định **`vi`**, `no_speech_thold=0.6`, bỏ segment im lặng / marker hallucination.
6. Text trim → (qua `convert_path`) NFC normalize → stdout.

```mermaid
flowchart LR
  A[mp3/wav/ogg] --> B[symphonia decode]
  B --> C[mono average]
  C --> D[rubato FFT 16kHz]
  D --> E[Whisper full lang=vi]
  E --> F[filter no_speech / markers]
  F --> G[text Markdown]
```

## LoadOnceCache (không có TTL theo thời gian)

- Slot: `Loading | Ready | Failed` + `Condvar` — 1 loader/key; waiter chờ, không load chồng.
- Capacity mặc định **2** (`FILECONV_WHISPER_CACHE_CAPACITY`); **không** expire theo thời gian (issue ghi “TTL/capacity” → thực tế chỉ có **capacity + LRU**).
- Key `WhisperModelKey` = canonical path + `use_gpu` / `flash_attn` / `gpu_device`. `threads` và `no_speech` **không** vào key (set trên `AudioEngine` sau khi lấy context).
- Evict chỉ drop entry trong map; `AudioEngine` còn giữ `Arc` thì model vẫn sống tới khi drop hết.

### LRU eviction kích hoạt khi nào?

Khi số slot **Ready** vượt capacity (default 2) sau khi publish model mới vào Ready — bỏ key Ready cũ nhất (LRU). Failed không tính Ready; Failed ở lại để caller retry (`Failed → Loading`).

## Resample — vì sao rubato FFT chứ không linear?

Whisper cần PCM f32 mono 16 kHz sạch. Linear dễ alias tần số cao khi down-sample. Rubato FFT (windowed anti-alias) chất lượng hơn cho tiếng Việt; code trim `output_delay` sau flush, lỗi DSP → `Err` (không bịa buffer silence).

## AC

- [x] Chạy audio convert thành công (có model; tiny + so sánh PhoWhisper)
- [x] Giải thích LoadOnceCache: Loading/Ready/Failed + capacity (không TTL thời gian) + khi nào LRU evict
- [x] Mô tả pipeline decode → resample 16kHz → whisper (`vi` + filter)
- [x] Trả lời: LRU eviction khi Ready > capacity; rubato FFT vì anti-alias / chất lượng hơn linear
- [x] Không train model, không đổi `no_speech` threshold, không PR/branch (read-only)
